pkg/smart_wallet/init_code.go

```diff
@@ -1,92 +1,47 @@
-	return len(byteCode) != 0, nil
-}
+func GetInitCode(smartWalletConfig Config, ownerAddress common.Address, index decimal.Decimal) ([]byte, error) {
+	var initCode []byte
+	var err error
 
-func GetAccountAddress(ctx context.Context, provider ethereum.ContractCaller, config Config, entryPointAddress, owner common.Address, index decimal.Decimal) (common.Address, error) {
-	initCode, err := GetInitCode(config, owner, index)
-	if err != nil {
-		return common.Address{}, fmt.Errorf("failed to build initCode middleware: %w", err)
+	switch typ := *smartWalletConfig.Type; typ {
+	case SimpleAccountType:
+		return nil, fmt.Errorf("%w: %s", ErrSmartWalletNotSupported, typ)
+	case BiconomyType: // not tested
+		initCode, err = GetBiconomyInitCode(ownerAddress, index, smartWalletConfig.Factory, smartWalletConfig.ECDSAValidator)
+	case KernelType:
+		initCode, err = GetKernelInitCode(ownerAddress, index, smartWalletConfig.Factory, smartWalletConfig.Logic, smartWalletConfig.ECDSAValidator)
+	default:
+		return nil, fmt.Errorf("unknown smart wallet type: %s", typ)
 	}
 
-	// calculate the smart wallet address that will be generated by the entry point
-	// See https://github.com/eth-infinitism/account-abstraction/blob/v0.6.0/contracts/core/EntryPoint.sol#L356
-	getSenderAddressData, err := entryPointABI.Pack("getSenderAddress", initCode)
 	if err != nil {
-		return common.Address{}, fmt.Errorf("failed to pack getSenderAddress data: %w", err)
-	}
-
-	msg := ethereum.CallMsg{
-		To:   &entryPointAddress,
-		Data: getSenderAddressData,
-	}
-
-	// this call must always revert (see EntryPoint contract), so we expect an error
-	_, err = provider.CallContract(ctx, msg, nil)
-	if err == nil {
-		panic(fmt.Errorf("'getSenderAddress' call returned no error, but expected one"))
-	}
-
-	var scError rpc.DataError
-	if ok := errors.As(err, &scError); !ok {
-		return common.Address{}, fmt.Errorf("unexpected error type '%T' containing message %w)", err, err)
-	}
-	errorData := scError.ErrorData().(string)
-
-	// check if the error signature is correct
-	if id := senderAddressResultError.ID.String(); errorData[0:10] != id[0:10] {
-		return common.Address{}, fmt.Errorf("'getSenderAddress' unexpected error signature: %s", errorData[0:10])
-	}
-
-	// check if the error data has the correct length
-	if len(errorData) < 74 {
-		return common.Address{}, fmt.Errorf("'getSenderAddress' revert data expected to have lenght of 74, but got: %d", len(errorData))
-	}
-
-	swAddress := common.HexToAddress(errorData[34:])
-	if swAddress == (common.Address{}) {
-		return common.Address{}, fmt.Errorf("'getSenderAddress' returned zero address")
+		return nil, fmt.Errorf("failed to get init code: %w", err)
 	}
 
-	return swAddress, nil
+	return initCode, nil
 }
 
-func GetInitCode(smartWalletConfig Config, ownerAddress common.Address, index decimal.Decimal) ([]byte, error) {
+func GetInitCodeFromFactoryCallData(smartWalletConfig Config, factoryCallData []byte) ([]byte, error) {
```

Comment:
I suggest adding godoc for this function since other public functions in this file have it.

Reply:
6982100528d7e2cfb6f19798fbceeaf3e24497b1